oshi-core/src/main/java/oshi/hardware/platform/linux/LinuxComputerSystem.java

```diff
@@ -44,48 +44,68 @@
     /** {@inheritDoc} */
     @Override
     public String getManufacturer() {
-        if (this.manufacturer == null && !queryManufacturerFromSysfs() && !queryManufacturerFromProcCpu()) {
-            this.manufacturer = Constants.UNKNOWN;
+        String localRef = this.manufacturer;
+        if (localRef == null) {
+            synchronized (this) {
+                localRef = this.manufacturer;
+                if (localRef == null) {
+                    if (!queryManufacturerFromSysfs() && !queryManufacturerFromProcCpu()) {
```

Comment:
The `manufacturer` field is declared in the super class `AbstractComputerSystem` but is written in subclasses. This violates encapsulation for no good reason. I am suggesting the following:

- Declare `AbstractComputerSystem.manufacturer` as `private`.
- Remove the `getManufacturer` method from all subclasses of `AbstractComputerSystem` and move it to `AbstractComputerSystem`.
- Introduce the `abstract protected String AbstractComputerSystem.manufacturer()` method specifying the logic of getting the value of `manufacturer`.
- Use the `manufacturer()` method in `AbstractComputerSystem` in order to calculate the value of `manufacturer`.
- Implement the `manufacturer()` method in all subclasses of `AbstractComputerSystem`. E.g. for `LinuxComputerSystem` the implementation would be
```java
protected final String manufacturer() {
  String result;
  if ((result = queryManufacturerFromSysfs()) == null &&
      (result = queryManufacturerFromProcCpu()) == null) {
    return Constants.UNKNOWN;
  }
  return result;
}
```
This would of course require to change `queryManufacturerFromSysfs()` and `queryManufacturerFromProcCpu()` methods so that they return manufacturer instead of writing to the `manufacturer` field. But such a change again would make the code much less error-prone and more readable, because we would have a single place where the field `manufacturer` is written, instead of having multiple places as it is now.

Reply:
Thanks for this suggestion, and I did attempt to do something similar to that.   However, the pattern breaks down (and possibly the earlier proposed memoizer pattern) because there is not a 1-to-1 relationship of value-fetching logic to value assignment.  For the `ComputerSystem` object, as an example, while Linux fetches the three values separately:
- Windows uses a WMI query that returns both Manufacturer and Model
- Mac uses the `system_profiler` command line to get both Model and Serial Number
- FreeBSD uses `dmidecode` command line to get Manufacturer, Model, and Serial Number
- Solaris uses the `smbios` command line that returns both Manufacturer and Model

The pattern above, fetching `manufacturer()` does not allow us to use the result to populate the `model` field, unless I mess around with arrays or other encapsulation.

Certainly I could fetch things individually, and in these cases the cost of repeating the commands is minimal, but there are many similar situations where a single command fetches values for multiple fields in a more expensive manner (usually WMI queries) and I'd like a (thread safe) way of populating all the fields at once.

Reply:
I see. Yeah, that may affect both the idea mentioned here (not very important) and the idea of using memoizer (important). I understand and agree with the reason you don't want to initialize the fields individually. However, there is at least one another approach that can be taken: move the fields initialized in groups out of `AbstractComputerSystem ` into subclasses.

You mentioned that `manufacturer` and `model` are initialized at once in `WindowsComputerSystem` via `queryManufacturerAndModel`. This means we may remove these fields from `AbstractComputerSystem` and have the following:
```java
final class WindowsComputerSystem extends AbstractComputerSystem {
  private final Supplier<ManufacturerAndModel> manufacturerAndModel = memoized(this:: queryManufacturerAndModel);

  public String getManufacturer() {
    return manufacturerAndModel.get().manufacturer;
  }

  public String getModel() {
    return manufacturerAndModel.get().model;
  }

  private ManufacturerAndModel queryManufacturerAndModel() {
    ...
    return new ManufacturerAndModel(...);
  }

  private static final class ManufacturerAndModel {
    private final String manufacturer;
    private final String model;

    private ManufacturerAndModel() {
      this.manufacturer = manufacturer;
      this.model = model;
    }
  }
}
```

Other concrete implementations of `ComputerSystem ` may do a similar thing. If this approach is taken, then:
- only the `baseboard ` and `firmware ` fields would stay in `AbstractComputerSystem`;
- `LinuxComputerSystem` would have to declare all three fields `manufacturer`, `model`, `serialNumber` because `AbstractComputerSystem` would not be declaring them anymore.

While this approach misses the opportunity of declaring the state in a single place (`AbstractComputerSystem`), it leaves us with the benefit of using memoizers. In other words we are effectively choosing between
- duplicating field declarations
- duplicating multi-threaded lazy initialization code.

Despite both options are not good, I think the former is less bad than the latter. And unfortunately, I can not see a way of solving both of the problems in this case.

Reply:
@stIncMale Thanks so much for taking the time to educate me!

I've updated the code with your suggestions:
- Used a Memoizer instead of double checked locking, implemented in SystemInfo, AbstractHardwareAbstrationLayer, and AbstractComputerSystem.  Working through the platform-specific ones, have only done the WindowsComputerSystem per your suggestion
- Re-established encapsulation

It's getting a bit more complex now with SolarisComputerSystem as one method call only returns two of the fields and another returns three, so it will take a little bit of thought... but I think I'm on a roll here!